# Introdução ao Python

## Características Básicas da Linguagem

### Operações matemáticas

#### Adição (+)

In [55]:
2+25

27

#### Subtração (-)

In [56]:
3-68

-65

#### Divisão (/) e (//)

A operação divisão sempre retorna um número de ponto flutuante

In [57]:
12/5

2.4

In [58]:
12//5

2

In [59]:
type(12/5)

float

#### Exponenciação (**)

In [60]:
2**3

8

#### Resto da divisão (%)

Serva para saber se um valor é divisível ou multiplo de outro

In [61]:
# Para um valor se divisível por outro, a operação resto da divisão será 0

# Criação da função que averigua e retorna se um valor x é multiplo de um valor y
def multiplo (x,y):
    if x%y == 0:
        print(x,'é múltiplo de', y)
    else:
        print(x,'não é múltiplo de', y)

In [62]:
multiplo(4,2)
multiplo(357,4)

4 é múltiplo de 2
357 não é múltiplo de 4


#### Expressões Matemáticas

In [63]:
2**2+4*4

20

In [64]:
10%2+3//10

0

In [65]:
5*(2+3)/2

12.5

In [66]:
2**3*4

32

#### A variável _

Quando o _ (underscore) aparece ele se refere ao último valor impresso. Ou seja, o último valor impresso é atribuído a ele

In [67]:
5*2

10

In [68]:
_+5

15

In [69]:
_/3

5.0

In [70]:
print('Hello World')

Hello World


In [71]:
print(_)

5.0


Como é possível observar a variável _ não acumula string

# Trabalhando com arrays Numpy

In [ ]:
#Instalação do pacote Numpy
%pip install numpy

In [73]:
#Importação do pacote Numpy
import numpy as np

In [74]:
anos = np.loadtxt("D:\\Python\\carros-anos.txt")
km = np.loadtxt("D:\\Python\\carros-km.txt")
valor = np.loadtxt("D:\\Python\\carros-valor.txt")
zero_km = np.loadtxt("D:\\Python\\carros-zero-km.txt")

In [75]:
dataset = np.column_stack((anos,km,valor,zero_km))
dataset

array([[  2003.  ,  44410.  ,  88078.64,      0.  ],
       [  1991.  ,   5712.  , 106161.94,      0.  ],
       [  1990.  ,  37123.  ,  72832.16,      0.  ],
       ...,
       [  2002.  ,  52637.  ,  68934.03,      0.  ],
       [  1996.  ,   7685.  , 122110.9 ,      0.  ],
       [  1992.  ,  50188.  ,  90381.47,      0.  ]])

In [76]:
dataset.shape

(258, 4)

# Cálculos estatísticos

### Média

np.mean() -> Retorna a média dos elementos do array ao longo do eixo especificado

In [77]:
# Média de todos os valores do dataset excluindo a última coluna que é a variável booleana
np.mean(dataset[:,0:3])

48489.14648578811

In [78]:
# Média de cada coluna do dataset excluindo a última coluna que é a variável booleana
np.mean(dataset[:,0:3],axis=0)

array([ 2007.51162791, 44499.41472868, 98960.51310078])

In [79]:
# Média cada linha do dataset, excluindo a última coluna que é a variável boolena
np.mean(dataset[:,0:3], axis=1)

array([44830.54666667, 37954.98      , 37315.05333333, 42189.35666667,
       40125.03333333, 36745.91      , 19488.06666667, 63972.81333333,
       73974.42333333, 38851.83      , 28836.19666667, 27992.84333333,
       68168.99666667, 18400.79333333, 56984.80666667, 59177.8       ,
       36526.53      , 30190.46333333, 33932.89666667, 39638.76666667,
       69780.79333333, 32194.46666667, 77867.61333333, 76878.21666667,
       41205.07      , 24462.03333333, 34386.25      , 35817.28      ,
       39298.42333333, 72268.23333333, 34289.06666667, 41477.39333333,
       26365.56333333, 33918.13333333, 50905.01      , 54959.84      ,
       17875.19666667, 41766.95666667, 25768.39333333, 34876.18      ,
       54475.06333333, 59333.43      , 29553.44666667, 75040.79666667,
       39881.91      , 22655.23333333, 52057.21666667, 31725.88333333,
       22240.03333333, 69567.49666667, 78571.96333333, 61600.27666667,
       69114.54333333, 58791.90333333, 31242.99666667, 51282.34333333,
      

In [80]:
# Caso queira fazer o cálculo da média de uma coluna específica, basta i9ndexar qual coluna você quer
# Média referente a coluna com a quilometragem dos carros
np.mean(dataset[:,1])

44499.41472868217

### Outras funções de cálculo

np.std -> retorna o desvio padrão do elementos do array ao longo do eixo especificado

np.sum ou ndarray.sum -> retorna a soma dos elementos do array ao longo do eixo específicado

In [81]:
# Cálculo dos desvios padrões das colunas do dataset excluindo a variável boolena
np.std(dataset[:,0:3], axis=0)

array([9.70703915e+00, 3.98598270e+04, 2.97541012e+04])

In [82]:
# Cálculo das somas das colunas do dataset excluindo a variável boolena
np.sum(dataset[:,0:3], axis=0)

array([  517938.        , 11480849.        , 25531812.37999999])

In [83]:
# Cálculo das somas das colunas do dataset excluindo a variável boolena
dataset[:,0:3].sum(axis=0)

array([  517938.        , 11480849.        , 25531812.37999999])

### Ainda é possível determinar a quantidade de carros novos e carros usados que temos

Se zero_km = 0 -> carro usado

Se zero_km = 1 -> carro novo

In [84]:
#Exemplo da aplicação das extruturas de repetição e condicional
carro_novo = 0
carro_usado = 0
for i in dataset[:,3]:
    if i == 1:
        carro_novo += 1
    else:
        carro_usado += 1

print('Temos um quantitativo de ',carro_novo, 'carros novos')
print('Temos um quantitativo de ',carro_usado, 'carros usados')

Temos um quantitativo de  61 carros novos
Temos um quantitativo de  197 carros usados


In [85]:
#Exemplo de aplicação da função sum
print('Temos um quantitativo de ',sum(dataset[:,3]==1), 'carros novos')
print('Temos um quantitativo de ',sum(dataset[:,3]==0), 'carros usados')

Temos um quantitativo de  61 carros novos
Temos um quantitativo de  197 carros usados


In [86]:
#Criação da coluna que vai conter os dados da quilometragem média para cada veículo
km_media = km/(2021-anos)
km_media

array([2.46722222e+03, 1.90400000e+02, 1.19751613e+03, 0.00000000e+00,
       1.71713333e+03, 1.19200000e+03, 0.00000000e+00, 6.46658333e+03,
       9.01790909e+03, 3.79780000e+03, 6.76789474e+02, 5.75142857e+02,
       4.48865000e+03, 0.00000000e+00, 3.45475000e+03, 2.31214000e+04,
       5.16100000e+03, 0.00000000e+00, 7.41720000e+03, 3.79325000e+03,
       5.32594737e+03, 0.00000000e+00, 3.63255556e+03, 3.43196667e+03,
       0.00000000e+00, 0.00000000e+00, 1.15900000e+03, 0.00000000e+00,
       3.09726316e+03, 3.14603333e+03, 1.25679167e+03, 2.05123077e+03,
       5.90666667e+03, 1.12693333e+03, 1.29548571e+04, 1.69134615e+03,
       0.00000000e+00, 0.00000000e+00, 4.60500000e+02, 0.00000000e+00,
       1.16768750e+04, 2.26455556e+03, 0.00000000e+00, 4.54221053e+03,
       0.00000000e+00, 5.13421053e+02, 4.11441176e+03, 1.19750000e+02,
       0.00000000e+00, 4.46383333e+03, 7.12950000e+03, 3.36640000e+03,
       3.22457143e+03, 6.26313333e+03, 3.21318182e+03, 3.00025926e+03,
      

In [87]:
carros_info = np.column_stack((dataset,km_media))
carros_info

array([[  2003.        ,  44410.        ,  88078.64      ,
             0.        ,   2467.22222222],
       [  1991.        ,   5712.        , 106161.94      ,
             0.        ,    190.4       ],
       [  1990.        ,  37123.        ,  72832.16      ,
             0.        ,   1197.51612903],
       ...,
       [  2002.        ,  52637.        ,  68934.03      ,
             0.        ,   2770.36842105],
       [  1996.        ,   7685.        , 122110.9       ,
             0.        ,    307.4       ],
       [  1992.        ,  50188.        ,  90381.47      ,
             0.        ,   1730.62068966]])

In [88]:
#Determinação do carro com maior quilometragem média
print('O carro que teve maior quilometragem média rodou em média',max(carros_info[:,4]),'km/ano')

O carro que teve maior quilometragem média rodou em média 29940.0 km/ano


Como podemos ver a biblioteca Numpy é muito útil para cálculos matemáticos, porém tem uma limitação que é o uso de apenas valores numéricos em sua operação. No cenário acima seria interessante saber qual modelo de carro tem a mior quilometragem rodada.

Sendo assim vamos agora conhecer um pouco outro pacote do Python que é o pandas.

In [ ]:
#Instalação do pacote Pandas
%pip install pandas

In [90]:
#Importação do pacote Pandas e nomeando ele como pd
import pandas as pd

# Criação de DataFrames

Como já temos alguns vetores com os dados que temos interesse, vamos criar um DataFrame com todos os dados acerca dos carros, agora incluindo o modelo de cada um

In [91]:
#Leitura de arquivo txt que está alocado em uma pasta local na máquina
nomes = open("D:\\Python\\carros-nomes.txt",'r')
nomes = pd.DataFrame(nomes)
nomes

0
0            Jetta Variant\n
1                   Passat\n
2                 Crossfox\n
3                      DS5\n
4         Aston Martin DB4\n
..                       ...
253           Phantom 2013\n
254  Cadillac Ciel concept\n
255             Classe GLK\n
256       Aston Martin DB5\n
257                  Macan\n

[258 rows x 1 columns]

In [92]:
#Criação de um Data Frame a partir de vários arrays numpy
#Essa função .T é para transpor a matriz, sem ela os arrays ficam nas linhas. Contudo o interessante é que cada vetor ocupe uma coluna, por isso devemos transpor a matriz
carros_df = pd.DataFrame((anos,km,km_media,zero_km,valor)).T
carros_df

0        1            2    3          4
0    2003.0  44410.0  2467.222222  0.0   88078.64
1    1991.0   5712.0   190.400000  0.0  106161.94
2    1990.0  37123.0  1197.516129  0.0   72832.16
3    2019.0      0.0     0.000000  1.0  124549.07
4    2006.0  25757.0  1717.133333  0.0   92612.10
..      ...      ...          ...  ...        ...
253  2014.0  27505.0  3929.285714  0.0   51759.58
254  1991.0  29981.0   999.366667  0.0   51667.06
255  2002.0  52637.0  2770.368421  0.0   68934.03
256  1996.0   7685.0   307.400000  0.0  122110.90
257  1992.0  50188.0  1730.620690  0.0   90381.47

[258 rows x 5 columns]

In [93]:
#Criação de uma nova coluna num Data Frame já existente
carros_df['Nomes'] = nomes
carros_df

0        1            2    3          4                    Nomes
0    2003.0  44410.0  2467.222222  0.0   88078.64          Jetta Variant\n
1    1991.0   5712.0   190.400000  0.0  106161.94                 Passat\n
2    1990.0  37123.0  1197.516129  0.0   72832.16               Crossfox\n
3    2019.0      0.0     0.000000  1.0  124549.07                    DS5\n
4    2006.0  25757.0  1717.133333  0.0   92612.10       Aston Martin DB4\n
..      ...      ...          ...  ...        ...                      ...
253  2014.0  27505.0  3929.285714  0.0   51759.58           Phantom 2013\n
254  1991.0  29981.0   999.366667  0.0   51667.06  Cadillac Ciel concept\n
255  2002.0  52637.0  2770.368421  0.0   68934.03             Classe GLK\n
256  1996.0   7685.0   307.400000  0.0  122110.90       Aston Martin DB5\n
257  1992.0  50188.0  1730.620690  0.0   90381.47                  Macan\n

[258 rows x 6 columns]

In [94]:
#Renomeando colunas de um Data Frame já existente
carros_df = carros_df.rename(columns={0:'Ano',1:'Quilometragem',2:'Quilometragem/Ano',3:'Novo/Usado',4:'Valor'})
carros_df

Ano  Quilometragem  Quilometragem/Ano  Novo/Usado      Valor  \
0    2003.0        44410.0        2467.222222         0.0   88078.64   
1    1991.0         5712.0         190.400000         0.0  106161.94   
2    1990.0        37123.0        1197.516129         0.0   72832.16   
3    2019.0            0.0           0.000000         1.0  124549.07   
4    2006.0        25757.0        1717.133333         0.0   92612.10   
..      ...            ...                ...         ...        ...   
253  2014.0        27505.0        3929.285714         0.0   51759.58   
254  1991.0        29981.0         999.366667         0.0   51667.06   
255  2002.0        52637.0        2770.368421         0.0   68934.03   
256  1996.0         7685.0         307.400000         0.0  122110.90   
257  1992.0        50188.0        1730.620690         0.0   90381.47   

                       Nomes  
0            Jetta Variant\n  
1                   Passat\n  
2                 Crossfox\n  
3                      DS5\n  
4         Aston Martin DB4\n  
..                       ...  
253           Phantom 2013\n  
254  Cadillac Ciel concept\n  
255             Classe GLK\n  
256       Aston Martin DB5\n  
257                  Macan\n  

[258 rows x 6 columns]

In [95]:
#Rearrajando as colunas por coveniência. Vale salientar que aqui é você quem escolhe a ordem das colunas como você julgar ficar mais organizado
carros_df = carros_df[['Nomes','Ano','Quilometragem','Quilometragem/Ano','Novo/Usado','Valor']]
carros_df.head(5)

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado  \
0     Jetta Variant\n  2003.0        44410.0        2467.222222         0.0   
1            Passat\n  1991.0         5712.0         190.400000         0.0   
2          Crossfox\n  1990.0        37123.0        1197.516129         0.0   
3               DS5\n  2019.0            0.0           0.000000         1.0   
4  Aston Martin DB4\n  2006.0        25757.0        1717.133333         0.0   

       Valor  
0   88078.64  
1  106161.94  
2   72832.16  
3  124549.07  
4   92612.10

In [96]:
#Seleção apenas dos carros novos, ou seja, que ainda não rodaram
carros_novos = carros_df[carros_df['Novo/Usado']==1]
carros_novos.head(5)

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado  \
3         DS5\n  2019.0            0.0                0.0         1.0   
6          A5\n  2019.0            0.0                0.0         1.0   
13         J5\n  2019.0            0.0                0.0         1.0   
17         A3\n  2019.0            0.0                0.0         1.0   
21  Serie 1 M\n  2019.0            0.0                0.0         1.0   

        Valor  
3   124549.07  
6    56445.20  
13   53183.38  
17   88552.39  
21   94564.40

In [97]:
carros_novos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61 entries, 3 to 248
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Nomes              61 non-null     object 
 1   Ano                61 non-null     float64
 2   Quilometragem      61 non-null     float64
 3   Quilometragem/Ano  61 non-null     float64
 4   Novo/Usado         61 non-null     float64
 5   Valor              61 non-null     float64
dtypes: float64(5), object(1)
memory usage: 3.3+ KB


A função info() traz algumas informações básicas sobre o nosso DataFrame. Podemos ver que temos 61 carros novos. E também é possível ver o tipo de variável que cada coluna tem, nesse caso temos variáveis do tipo object e float64.

In [98]:
carros_novos.describe()

Ano  Quilometragem  Quilometragem/Ano  Novo/Usado          Valor
count    61.0           61.0               61.0        61.0      61.000000
mean   2019.0            0.0                0.0         1.0  100423.811803
std       0.0            0.0                0.0         0.0   30156.009638
min    2019.0            0.0                0.0         1.0   50742.100000
25%    2019.0            0.0                0.0         1.0   71367.100000
50%    2019.0            0.0                0.0         1.0  102609.540000
75%    2019.0            0.0                0.0         1.0  124549.070000
max    2019.0            0.0                0.0         1.0  149157.170000

A função describe() traz informações estatísticas iniciais. Por exemplo, podemos ver que a primeira linha apresentada (count) é a contagem de elementos que tem em cada coluna. A linha das médias (mean) mostra a média dos valores de cada coluna. A linha std moostra o desvio padrão dos valores contidos na coluna. As funções min e max retornam os valos mínimo e máximo, respectivamente, de cada coluna. As colunas 25%, 50% e 75% mostram os quartis da distribuição de valores de cada coluna. Repare que foram usadas somente as colunas que tem valores numéricos, as colunas com tipos object não são usadas na função describe().
Então podemos confirmar que temos 61 carros e que de fato eles são novos, pois as colunas Quilometragem e Quuilometragem/Ano estão com valores 0. Ainda podemos ver que todos os carros datam do ano de 2019 e que o carro novo mais caro custa R$ 149157,17 enquanto que o carro novo mais barato custa R$ 50742,10.

In [99]:
#Informações sobre o carro novo mais caro
carros_novos[carros_novos['Valor']==max(carros_novos['Valor'])][['Nomes','Ano','Valor']]

Nomes     Ano      Valor
117  Lamborghini Ferruccio\n  2019.0  149157.17

In [100]:
#Informações sobre o carro novo mais barato
carros_novos[carros_novos['Valor']==min(carros_novos['Valor'])][['Nomes','Ano','Valor']]

Nomes     Ano    Valor
79  Livina\n  2019.0  50742.1

In [101]:
carros_usados = carros_df[carros_df['Novo/Usado']==0]
carros_usados.head(5)

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado  \
0     Jetta Variant\n  2003.0        44410.0        2467.222222         0.0   
1            Passat\n  1991.0         5712.0         190.400000         0.0   
2          Crossfox\n  1990.0        37123.0        1197.516129         0.0   
4  Aston Martin DB4\n  2006.0        25757.0        1717.133333         0.0   
5     Palio Weekend\n  2012.0        10728.0        1192.000000         0.0   

       Valor  
0   88078.64  
1  106161.94  
2   72832.16  
4   92612.10  
5   97497.73

In [102]:
carros_usados.describe()

Ano  Quilometragem  Quilometragem/Ano  Novo/Usado  \
count   197.000000     197.000000         197.000000       197.0   
mean   2003.954315   58278.421320        4803.804629         0.0   
std       8.380793   35836.733259        5401.530370         0.0   
min    1990.000000     107.000000           3.566667         0.0   
25%    1996.000000   27505.000000        1706.333333         0.0   
50%    2004.000000   55083.000000        3405.080000         0.0   
75%    2011.000000   90495.000000        5402.352941         0.0   
max    2018.000000  119945.000000       29940.000000         0.0   

               Valor  
count     197.000000  
mean    98507.410457  
std     29767.374513  
min     51667.060000  
25%     70709.340000  
50%     96369.040000  
75%    124661.380000  
max    149489.920000

Podemos ver que temos um total de 197 carros na agência. O carro mais antigo data de 1990 e o mais recente de 2018. Em média os carros usados rodam 4803,8 km/ano e tem um valor médio de R$ 98507,41. Sendo o carro usado mais caro num valor de R$ 149489,92 e o mais barato num valor de R$ 51667,06.

In [103]:
#Informações sobre o carro usado mais caro
carros_usados[carros_usados['Valor']==max(carros_usados['Valor'])][['Nomes','Ano','Quilometragem','Quilometragem/Ano','Valor']]

Nomes     Ano  Quilometragem  Quilometragem/Ano      Valor
128  ASX\n  2003.0        61672.0        3426.222222  149489.92

In [104]:
#Informações sobre o carro usado mais barato
carros_usados[carros_usados['Valor']==min(carros_usados['Valor'])][['Nomes','Ano','Quilometragem','Quilometragem/Ano','Valor']]

Nomes     Ano  Quilometragem  Quilometragem/Ano  \
254  Cadillac Ciel concept\n  1991.0        29981.0         999.366667   

        Valor  
254  51667.06

In [105]:
#Informação do carro mais caro da agência
carros_df[carros_df['Valor']==max([max(carros_usados['Valor']),max(carros_novos['Valor'])])]

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado      Valor
128  ASX\n  2003.0        61672.0        3426.222222         0.0  149489.92

In [106]:
#Informação do carro mais barato da agência
carros_df[carros_df['Valor']==min([min(carros_usados['Valor']),min(carros_novos['Valor'])])]

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado    Valor
79  Livina\n  2019.0            0.0                0.0         1.0  50742.1

In [107]:
#Informação do carro mais conservado, ou seja, o carro usado com menor quilometragem/ano
carros_usados[carros_usados['Quilometragem/Ano']==min(carros_usados['Quilometragem/Ano'])]

Nomes     Ano  Quilometragem  Quilometragem/Ano  Novo/Usado      Valor
227  X1\n  1991.0          107.0           3.566667         0.0  102968.76